### PR TITLE
#1173 inotify for configuration management  

### LINKED ISSUE #1173: c8y-configuration-plugin to detect local updates to its config file
When the `c8y-configuration-plugin.toml` file is updated locally from the device, the plugin doesn't reload its contents as done by the `c8y-log-plugin`. The same file watcher mechanism implemented for `c8y-log-plugin` must be implemented for the `c8y-configuration-plugin` as well. Whenever the config file is updated, either locally or via cloud, the updated list of config types must be sent to Cumulocity via the SmartREST 119 message.

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes

- Added a new function in tedge_test_utils that helps create inotify event streams.
- c8y log plugin and c8y configuration plugin now re-use this inotify function.
- c8y configuration toml file is updated on CLOSE_WRITE. This update happens if you change the file locally AND if you upload a config file from the cloud.  

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)

## Paste Link to the issue
Closes #1173

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/common/tedge_utils/src/inotify.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+pub use inotify_crate::WatchMask;
+use inotify_crate::{EventStream, Inotify};
+use std::path::Path;
+
+pub fn inofity_file_watch_stream(
+    config_file: &Path,
+    watch_flags: WatchMask,
+) -> Result<EventStream<[u8; 1024]>, anyhow::Error> {
+    let buffer = [0; 1024];
+    let mut inotify = Inotify::init().expect("Error while initializing inotify instance");
```
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/inotify.rs:11]: What can cause this error? Don't forget that `expect` panics !

File: crates/common/tedge_utils/src/inotify.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+    inotify
+        .add_watch(config_file, watch_flags)
+        .expect("Failed to add file watch");
```
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/inotify.rs:15]: This `expect` must be replaced by a proper error handler. The file might be non-accessible or the path might have a typo.

File: crates/common/tedge_utils/src/inotify.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+pub use inotify_crate::EventMask;
+pub use inotify_crate::WatchMask;
+use inotify_crate::{EventStream, Inotify};
+use std::path::Path;
+
+pub fn inofity_file_watch_stream(
+    config_file: &Path,
```
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/inotify.rs:7]: `path` would be a better name: this can be a directory or a file.

I would also fix the function name.

File: plugins/c8y_log_plugin/src/main.rs [hunk 4]
```diff
@@ -134,16 +120,13 @@ async fn run(
...
-                            plugin_config = handle_dynamic_log_type_update(mqtt_client, config_file).await?;
-                        }
-                        _ => {}
+            Some(Ok(event)) = inotify_stream.next() => {
+                match event.mask {
+                    EventMask::CLOSE_WRITE => {
+                        plugin_config = handle_dynamic_log_type_update(mqtt_client, config_file).await?;
                     }
+                    _ => {}
                 }
```
Comment (didier-wenzek) [inline plugins/c8y_log_plugin/src/main.rs:129]: This is were you idea of adding a `filter_map` to the `inotify` stream can really shine.

1) To avoid this cascade of matches, first on ` Some(Ok(event))` and then on the `event.mask`.
2) To avoid this C-like enums: `EventMask::CLOSE_WRITE` and `WatchMask::CLOSE_WRITE`
3) And why not to use a single enum for `EventMask` and `WatchMask`.

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 5]
```diff
@@ -134,65 +135,65 @@ async fn run(
...
+                                    )
+                                    .await?
+                                }
+                                _ => {
+                                    // Ignore operation messages not meant for this plugin
+                                }
+                            }
+                        }
+                    } else {
+                        return Ok(());
```
Comment (didier-wenzek) [inline plugins/c8y_configuration_plugin/src/main.rs:180]: By returning here, the plugin stops when receiving a non-utf8 message. An error should be displayed and the program must continue.

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 6]
```diff
@@ -134,65 +135,65 @@ async fn run(
...
+                        // Reload the plugin config file
+                        plugin_config = PluginConfig::new(config_file_path);
+                        // Resend the supported config types
+                        let msg = plugin_config.to_supported_config_types_message()?;
+                        mqtt_client.published.send(msg).await?;
                 }
-            };
-
-            if let Err(err) = result {
-                error!("Handling of operation: '{payload}' failed with {err}");
```
Comment (didier-wenzek) [inline plugins/c8y_configuration_plugin/src/main.rs:188]: Where is traced this error now?

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 7]
```diff
@@ -208,7 +209,7 @@ fn create_operation_files(config_dir: &str) -> Result<(), anyhow::Error> {
         &format!("{config_dir}/c8y/c8y-configuration-plugin.toml"),
         "root",
         "root",
-        0o644,
+        0o775,
```
Comment (didier-wenzek) [inline plugins/c8y_configuration_plugin/src/main.rs:212]: Why this change ? There is no point to have this file executable.

Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:212]: Inotify needs this to watch the file content

File: crates/common/tedge_utils/src/inotify.rs (hunk 1 above)
Comment (initard) [reply crates/common/tedge_utils/src/inotify.rs:11]: It can be caused by: 
1. ~invalid value specified in flags~ (not relevant as the crate abstracts this away from the user)
2. maximum number of inotify instances has been reached
3. maximum system limit of file descriptors 
4. insufficient kernel memory available

See [this](https://linux.die.net/man/2/inotify_init1).

File: crates/common/tedge_utils/Cargo.toml [hunk 8]
```diff
@@ -15,7 +15,8 @@ logging = ["tracing", "tracing-subscriber"]
 [dependencies]
 anyhow = "1.0"
 futures = "0.3"
-nix = "0.23.1"
+inotify_crate = { version = "0.10", package = "inotify" }
```
Comment (rina23q) [inline crates/common/tedge_utils/Cargo.toml:18]: Is the used crate https://crates.io/crates/inotify? Why not the one below?
```
inotify = "0.10"
```

**Review by rina23q — COMMENTED** [review]
Some comments.

File: plugins/c8y_configuration_plugin/src/main.rs (hunk 7 above)
Comment (didier-wenzek) [reply plugins/c8y_configuration_plugin/src/main.rs:212]: This is definitely not the case. To watch a file one just needs to be able to read it.

That said you might experience an issue if you edit the file using an editor which rename a temporary file instead of opening it for write. Since you are watching `WatchMask::CLOSE_WRITE on a specific file, those events will never be observed because the watched file is moved.

=> You have to watch the directory containing the configuration file.

File: crates/common/tedge_utils/Cargo.toml (hunk 8 above)
Comment (initard) [reply crates/common/tedge_utils/Cargo.toml:18]: That's right, the [nix crate also has inotify](https://docs.rs/nix/latest/nix/sys/inotify/index.html) but the API is slightly different.

However, I notice now that as of nix 0.24 the inotify is a feature, meaning that the namespace will not overlap anymore. I will update nix to 0.24 and use inotify 0.10 as @rina23q  shows above

File: plugins/c8y_configuration_plugin/src/main.rs (hunk 5 above)
Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:183]: This won't show as outdated but I made some, please check @rina23q

File: plugins/c8y_configuration_plugin/src/main.rs (hunk 7 above)
Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:211]: I see what you mean... I was not using inotify correctly

File: plugins/c8y_configuration_plugin/src/main.rs (hunk 5 above)
Comment (rina23q) [reply plugins/c8y_configuration_plugin/src/main.rs:183]: I still don't get it why here should return `Ok(())`?

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 9]
```diff
@@ -134,65 +134,65 @@ async fn run(
...
-                error!("Handling of operation: '{payload}' failed with {err}");
+                _ => {}
             }
         }
+        }
     }
-
-    mqtt_client.close().await;
-
-    Ok(())
```
Comment (rina23q) [inline plugins/c8y_configuration_plugin/src/main.rs:195]: Why these two lines were removed?

File: plugins/c8y_configuration_plugin/src/main.rs (hunk 5 above)
Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:183]: If the mqtt connection is closed it will return `None`. The `else` block will pick it up and close the main function
```rust
if let Some(message) = message {
   ...
} else {
  // message is None and the connection has been closed
  return Ok(());
}

```

File: plugins/c8y_configuration_plugin/src/main.rs (hunk 9 above)
Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:195]: I believe this is addressed in the [else above](https://github.com/thin-edge/thin-edge.io/pull/1187/files/929f1d95970b050f4b8232f162dcf8f624bd08a1#r897962089), no?

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 10]
```diff
@@ -134,65 +134,65 @@ async fn run(
...
+                        }}
+                    } else {
+                        // message is None and the connection has been closed
+                        return Ok(());
                     }
+            }
+        Some(Ok(event)) = inotify_stream.next() => {
+            match event.mask {
+                EventMask::CLOSE_WRITE => {
+                    plugin_config = handle_dynamic_config_update(mqtt_client, config_file_path).await?;
```
Comment (albinsuresh) [inline plugins/c8y_configuration_plugin/src/main.rs:189]: Isn't that `?` call a bit risky? Leading to this daemon crashing if the config file is malformed or if the message publish fails?
